feat: implement GCP PubSub source (#1)
numaproj-contrib/gcp-pub-sub-source-go: README.md

# GCP Pub/Sub Source for Numaflow

The GCP Pub/Sub Source is a custom source for [Numaflow](https://numaflow.numaproj.io/) that integrates Google Cloud Pub/Sub into Numaflow pipelines.

## Quick Start
This guide will walk you through setting up a GCP Pub/Sub source in a Numaflow pipeline.

### Prerequisites
- [Numaflow installed on your Kubernetes cluster](https://numaflow.numaproj.io/quick-start/).
- Access to a Google Cloud Platform project with Pub/Sub enabled.

### Step-by-Step Guide

1. **Create a GCP Pub/Sub Topic and Subscription**: Use the Google Cloud Console or `gcloud` CLI to create a new topic and subscription.

> 2. **Set Up Environment Variables**: Define environment variables such as `PROJECT_ID`, `TOPIC_ID`, `SUBSCRIPTION_ID`, and `PUBSUB_EMULATOR_HOST` in your Kubernetes deployment. These variables are crucial for configuring the Pub/Sub source within your pipeline.

Reviewer: Please also add `MAX_EXTENSION_PERIOD`, let's make the list complete. Please also add brief introductions to each of the fields. README is the single place for users to explore how to use this udsource so let's make it a good read.

# GCP Pub/Sub Source for Numaflow

The GCP Pub/Sub Source is a custom source for [Numaflow](https://numaflow.numaproj.io/) that integrates Google Cloud Pub/Sub into Numaflow pipelines.

## Quick Start
This guide will walk you through setting up a GCP Pub/Sub source in a Numaflow pipeline.

### Prerequisites
- [Numaflow installed on your Kubernetes cluster](https://numaflow.numaproj.io/quick-start/).
- Access to a Google Cloud Platform project with Pub/Sub enabled.

### Step-by-Step Guide

1. **Create a GCP Pub/Sub Topic and Subscription**: Use the Google Cloud Console or `gcloud` CLI to create a new topic and subscription.

2. **Set Up Environment Variables**: Define environment variables such as `PROJECT_ID`, `TOPIC_ID`, `SUBSCRIPTION_ID`, `MAX_EXTENSION_PERIOD` and `PUBSUB_EMULATOR_HOST` in your Kubernetes deployment. These variables are crucial for configuring the Pub/Sub source within your pipeline.

>    PROJECT_ID: This is the unique identifier for your Google Cloud Project. Every resource in Google Cloud belongs to a project. This ID is used by the Pub/Sub client to know which project's Pub/Sub resources it should interact with.

Reviewer: ```suggestion
PROJECT_ID: This is the unique identifier for your Google Cloud Project. Every resource in Google Cloud belongs to a project. This ID is used by the Pub/Sub client to know which project's Pub/Sub resources it should interact with.
```

              - name: PUBSUB_EMULATOR_HOST
                value: "gcloud-pubsub.numaflow-system.svc.cluster.local:8681"
    - name: log-sink
      sink:
        log: {}
  edges:
    - from: pubsub-source
      to: log-sink
```

Then apply it to your cluster:
```bash
kubectl apply -f pubsub-source-pipeline.yaml
```

#### 3. Verify the Pipeline

Check if the pipeline is running:
```bash
kubectl get pipeline pubsub-source-pipeline
```

#### 4. Publish Messages to the GCP Pub/Sub Topic

Use the Google Cloud Console or `gcloud` CLI to publish messages to your Pub/Sub topic.

#### 5. Verify the Log Sink

Check the logs of the `log-sink` vertex to see if it received the message from Pub/Sub:
```bash
kubectl logs <log-sink-pod-name>
```

You should see output similar to:
```
Message received from Pub/Sub
```

#### 6. Clean Up

To delete the Numaflow pipeline:
```bash
kubectl delete pipeline pubsub-source-pipeline
```

To delete the Pub/Sub topic and subscription:
```bash
# Replace <topic-id> and <subscription-id> with your topic and subscription IDs
gcloud pubsub topics delete <topic-id>
gcloud pubsub subscriptions delete <subscription-id>

> ```

Reviewer: nit: can we make sections 1 and 6 consistent for Pub/Sub creation/deletion? either all detailed with command line or brief.

- Access to Google Cloud Platform project with Pub/Sub enabled.

**Step-by-Step Guide:**

1. **Create a GCP Pub/Sub Topic and Subscription:**
   - Use the Google Cloud Console or `gcloud` CLI to create a new Pub/Sub topic and subscription.

2. **Deploy a Numaflow Pipeline with GCP Pub/Sub Source:**
   - Prepare a Kubernetes manifest file, let’s say `pubsub-source-pipeline.yaml`. Here's a template to get you started:

```yaml
apiVersion: numaflow.numaproj.io/v1alpha1
kind: Pipeline
metadata:
  name: pubsub-source-pipeline
spec:
  vertices:
    - name: pubsub-source
      source:
        udsource:
          container:
            image: "quay.io/numaio/numaflow-go/gcp-pubsub-source-go:latest"
            env:
              - name: PROJECT_ID
                value: "your-gcp-project-id"
              - name: TOPIC_ID
                value: "your-pubsub-topic-id"
              - name: SUBSCRIPTION_ID
                value: "your-pubsub-subscription-id"
              - name: PUBSUB_EMULATOR_HOST
                value: "gcloud-pubsub.numaflow-system.svc.cluster.local:8681"
    - name: log-sink
      sink:
        log: {}
  edges:
    - from: pubsub-source
      to: log-sink
    
```

> 

Reviewer: ```suggestion
```

- Modify the environment variables as per your Google Cloud project configuration.
- Apply the configuration to your cluster by running:

  ```bash
  kubectl apply -f pubsub-source-pipeline.yaml

> By following these steps, you'll have a GCP Pub/Sub topic and subscription as sources for your Numaflow pipeline running in Kubernetes. Messages from Pub/Sub will be fetched by the pipeline and passed to the log sink for further processing.

Reviewer: ```suggestion

By following these steps, you'll have a GCP Pub/Sub topic and subscription as sources for your Numaflow pipeline running in Kubernetes. Messages from Pub/Sub will be fetched by the pipeline and passed to the log sink for printing out.
```

## Using Environment Variables to Specify the GCP Pub/Sub Source Configuration

To configure the GCP Pub/Sub source in your Numaflow pipeline using environment variables, you can follow this process:

**Using Environment Variables for GCP Pub/Sub Source Configuration:**

1. **Set Up Environment Variables:**
   - Define environment variables on your Kubernetes deployment to hold the necessary GCP Pub/Sub configuration, such as the project ID, topic ID, subscription ID, and Pub/Sub emulator host.
   - For example, set `PROJECT_ID`, `TOPIC_ID`, `SUBSCRIPTION_ID`, and `PUBSUB_EMULATOR_HOST` as environment variables in your pipeline deployment.

2. **Modify Your Pipeline Configuration:**

>    - In your Numaflow pipeline definition (`pubsub-source-pipeline.yaml`), you will reference these environment variables. Ensure your source container in the pipeline specification is configured to use these variables.

Reviewer: I think 1 and 2 are a bit redundant, can we merge them as one paragraph?

   Apply it with `kubectl apply -f pubsub-source-pipeline.yaml`.

4. **Verify the Pipeline**: Check its status with `kubectl get pipeline pubsub-source-pipeline`.

5. **Publish Messages to the GCP Pub/Sub Topic**: Use the `gcloud` CLI to publish messages:

   ```bash
   gcloud pubsub topics publish <your-pubsub-topic-id> --message "Your message here"
   ```

6. **Verify the Log Sink**: Check the logs of the `log-sink` vertex with `kubectl logs <log-sink-pod-name>` to confirm message reception.

7. **Clean Up**:
    - Delete the pipeline: `kubectl delete pipeline pubsub-source-pipeline`.
    - Delete the Pub/Sub topic and subscription:

      ```bash
      gcloud pubsub topics delete <topic-id>
      gcloud pubsub subscriptions delete <subscription-id>
      ```

## Debugging

Enable debugging by setting `DEBUG=true` in the GCP Pub/Sub source container's environment variables. For further details, see the [Numaflow Debugging Guide](https://numaflow.numaproj.io/development/debugging/).

## Additional Resources

For more information, visit the [Numaflow Documentation](https://numaflow.numaproj.io/) and [Google Cloud Pub/Sub Documentation](https://cloud.google.com/pubsub/docs).

> ed additions and details, ensuring a comprehensive and clear guide for setting up a GCP Pub/Sub source in a Numaflow pipeline.

Reviewer: Should we remove this extra line?
